posthog/hogql/database/schema/util/session_where_clause_extractor.py

```diff
@@ -0,0 +1,329 @@
+        else:
+            return ast.Or(exprs=filtered)
+
+    def visit_alias(self, node: ast.Alias) -> ast.Expr:
+        return self.visit(node.expr)
+
+
+def is_time_or_interval_constant(expr: ast.Expr) -> bool:
+    return IsTimeOrIntervalConstantVisitor().visit(expr)
+
+
+class IsTimeOrIntervalConstantVisitor(Visitor):
+    def visit_constant(self, node: ast.Constant) -> bool:
+        return True
+
+    def visit_compare_operation(self, node: ast.CompareOperation) -> bool:
+        return self.visit(node.left) and self.visit(node.right)
+
+    def visit_arithmetic_operation(self, node: ast.ArithmeticOperation) -> bool:
+        return self.visit(node.left) and self.visit(node.right)
+
+    def visit_call(self, node: ast.Call) -> bool:
+        # some functions just return a constant
+        if node.name in ["today", "now"]:
+            return True
+        # some functions return a constant if the first argument is a constant
+        if node.name in [
+            "parseDateTime64BestEffortOrNull",
+            "toDateTime",
+            "toTimeZone",
+            "assumeNotNull",
+            "toIntervalYear",
+            "toIntervalMonth",
+            "toIntervalWeek",
+            "toIntervalDay",
+            "toIntervalHour",
+            "toIntervalMinute",
+            "toIntervalSecond",
+            "toStartOfDay",
+            "toStartOfWeek",
+            "toStartOfMonth",
+            "toStartOfQuarter",
+            "toStartOfYear",
+        ]:
+            return self.visit(node.args[0])
+
+        if node.name in ["minus", "add"]:
+            return all(self.visit(arg) for arg in node.args)
+
+        # otherwise we don't know, so return False
+        return False
+
+    def visit_field(self, node: ast.Field) -> bool:
+        return False
+
+    def visit_and(self, node: ast.And) -> bool:
+        return False
+
+    def visit_or(self, node: ast.Or) -> bool:
+        return False
+
+    def visit_not(self, node: ast.Not) -> bool:
+        return False
+
+    def visit_placeholder(self, node: ast.Placeholder) -> bool:
+        raise Exception()
+
+    def visit_alias(self, node: ast.Alias) -> bool:
+        return self.visit(node.expr)
+
+
+def is_simple_timestamp_field_expression(expr: ast.Expr) -> bool:
+    return IsSimpleTimestampFieldExpressionVisitor().visit(expr)
+
+
+class IsSimpleTimestampFieldExpressionVisitor(Visitor):
+    def visit_constant(self, node: ast.Constant) -> bool:
+        return False
+
+    def visit_field(self, node: ast.Field) -> bool:
+        # this is quite leaky, as it doesn't handle aliases, but will handle all of posthog's hogql queries
+        return (
+            node.chain == ["min_timestamp"]
+            or node.chain == ["sessions", "min_timestamp"]
+            or node.chain == ["s", "min_timestamp"]
+            or node.chain == ["timestamp"]
+            or node.chain == ["events", "timestamp"]
+            or node.chain == ["e", "timestamp"]
```

Comment:
This could be rewritten as looking at the last element of the chain, reducing the need to account for aliases. Even better though, you could use the node `Type` to get info about what the underlying field being accessed actually is. e.g. it'll give you the underlying type (such as a `DateTime`) and field name on the table it's accessing. Looks like you just wanna check whether the field is either the `timestamp` field on the `Sessions` or `Events` table, maybe? e.g. this does something similar by using the types - https://github.com/PostHog/posthog/blob/tom/zendesk/posthog/hogql/database/schema/event_sessions.py#L88-L124